src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -0,0 +1,421 @@
+(* All rights reserved.                                                   *)
+(* This source code is licensed under the GNU Affero General Public       *)
+(* License version 3 found in the LICENSE.md file in the root directory   *)
+(* of this source tree.                                                   *)
+(*                                                                        *)
+(**************************************************************************)
+
+module NEL = Cobol_common.Basics.NEL
+
+let cmlyname = ref None
+let external_tokens = ref ""
+
+let usage_msg = Fmt.str "%s [OPTIONS] file.cmly" Sys.argv.(0)
+let anon str = match !cmlyname with
+  | None -> cmlyname := Some str
+  | Some _ -> raise @@ Arg.Bad "Only one anonymous argument may be given"
+
+let () =
+  Arg.parse
+    Arg.[
+      ("--external-tokens", Set_string external_tokens,
+       "<module> Import token type definition from <module>");
+    ]
+    anon usage_msg
+
+let cmlyname = match !cmlyname with
+  | None | Some "" -> Fmt.epr "%s@." usage_msg; exit 1
+  | Some s -> s
+
+(* --- UTILS --- *)
+
+let inv f a b = f b a
+
+let pp_brackets pp =
+  Fmt.(any "[" ++ pp ++ any "]")
+
+let pp_pair pp_a pp_b = fun ppf (a, b) ->
+    Fmt.pf ppf "(%a,%a)" pp_a a pp_b b
+
+let pp_list pp = Fmt.list ~sep:(Fmt.any ";") pp
+
+let rec pp_nel pp ppf = function
+  | NEL.One v -> Fmt.pf ppf "One %a" pp v
+  | v::nel -> Fmt.pf ppf "%a::" pp v; pp_nel pp ppf nel
+
+let pp_match_cases ppf pp_key pp_value default l =
+  List.iter (fun (keys, value) ->
+    Fmt.pf ppf "  | %a -> %a\n"
+      Fmt.(list ~sep:(any " | ") pp_key) keys
+      pp_value value) l;
+  Fmt.pf ppf "  | _ -> %s\n" default
+
+(* --- *)
+
+include MenhirSdk.Cmly_read.Read (struct let filename = cmlyname end)
+
+type completion_entry =
+  | K of terminal NEL.t
+  | Custom of string
+(* [@@deriving ord] *)
+
+let pp_completion_entry: completion_entry Fmt.t = fun ppf -> function
+  | K list -> Fmt.pf ppf "K (%a)" (pp_nel Print.terminal) list
+  | Custom custom_type -> Fmt.pf ppf "%s" custom_type
+
+let is_valid_for_comp: terminal -> bool = fun term ->
+  Terminal.kind term == `REGULAR &&
+  Terminal.attributes term |>
+  List.exists (fun attrib ->
+    Attribute.has_label "keyword" attrib ||
+    Attribute.has_label "completion" attrib ||
+    Attribute.has_label "keyword.combined" attrib)
+
+let terminal_filter_map: terminal -> completion_entry option = fun term ->
+  if is_valid_for_comp term then Some (K (NEL.One term)) else None
+
+let nonterminal_filter_map: nonterminal -> completion_entry option = fun nonterm ->
+  Nonterminal.attributes nonterm |>
+  List.find_opt (Attribute.has_label "completion") |>
+  Option.map Attribute.payload |>
+  inv Option.bind (fun s -> Some (Custom s))
+
+let pp_xsymbol_of_nt ppf nonterminal =
+  Fmt.pf ppf "X(N N_%s)" (Nonterminal.mangled_name nonterminal)
+
+(* --- *)
+
+module SuperSet (OrderedType : Set.OrderedType) = struct
```

Comment:
Why "Super"? It's just a regular `Make_set`.

- Ok for `add_option`;
- `add_list` should be removed and `Set.union s (Set.of_list l)` should be used instead;
- `from_list_map` is a bit far-fetched with its `add` labeled argument…